### ISSUE TITLE
Help wanted --- Selection logic problem

### ISSUE THREAD

**Issue by bclothier** [rubberduck-vba/Rubberduck#3581]
Given the input code:
```
Option Compare Database
Option Explicit

Public Sub Testit()
    Dim var1 As Long
    Dim var2 As Long
    Dim var3 As String
    
    Debug.Print var1
    Debug.Print var2
    MsgBox var3
End Sub
```

and given the selection on line 9 and 10, including the indention which evaluates to `{Main.zzzTest L9C1 - L10C21}`

and given the reference to the `var1` being located at selection `{L9C17 - L9C21}`

it should be obvious that the `IdentifierReference.IsSelected` should return true. But it does not....

The [current logic](https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.VBEEditor/Selection.cs#L35) in `Selection.cs` for the method `ContainsFirstCharacter` was converting the selection `{Main.zzzTest L9C1 - L10C21}` to `{L9C1}`, which fell outside the `var1`'s reference at the `{L9C17}`. Subsequently, it was revised accordingly:
```
public bool ContainsFirstCharacter(Selection selection)
{
    return Contains(new Selection(selection.StartLine, selection.StartColumn, selection.EndLine, selection.EndColumn));
}
```

This caused 2 unit tests to fail which was considered acceptable on the presumption that a bug was uncovered.

However, this did not fully solve the problem because when we enter the `Contains` on the line 40, we should be sent to this case:
```
            // starts on same line:
            if (selection.StartLine == StartLine && selection.StartColumn >= StartColumn)
            {
                return selection.EndLine < EndLine || 
                    (selection.EndLine == EndLine && selection.EndColumn <= EndColumn);
            }
```
Both `selection` which represents `{L9C1 - L10C21}` and the instance itself representing `{L9C17 - L9C21}` has same starting line so the first condition is satisfied.

However, the `selection.StartColumn` evaluates to `1` whereas the `StartColumn` evaluates to `17`, which is false. That is incorrect because the selection *does* encompass the instance.

The obvious thing to do was to flip the expression so it's `selection.StartColumn <= StartColumn`... but doing that yields 50+ failed unit tests which is unacceptably high. I'm hoping it's just *me* who has horribly misunderstood the code and thus my request for help reviewing the proposed code change.

For reference, the original calling code which yield the incorrect results described:
```
SourceVariables = State.DeclarationFinder.UserDeclarations(DeclarationType.Variable)
                .Where(d => d.IsSelected(Selection) ||
                            d.References.Any(r => r.IsSelected(Selection)));
```

In contrast, this calling code gives the correct output plus 2 false positives, due to not comparing the `ComponentName`
```
            SourceVariables = State.DeclarationFinder.UserDeclarations(DeclarationType.Variable)
                .Where(d => Selection.Selection.Contains(d.Selection) ||
                            d.References.Any(r => Selection.Selection.Contains(r.Selection)));
```

**Comment by bclothier** [comment]
Thinking about it more, it might be that the `ContainsFirstCharacter` is flawed and should be *reversing* the call, collapsing the current instance in a new `Selection` instance, then pass that to the `selection.Contains`, rather than to the instance's `Contains` selection. After all when we say `a` contains `b`, `a` should be the user's selection and the `b` should be the target but in the code above, it seems to be reversed.

**Comment by MDoerner** [comment]
I think the main confusion here is about what `IsSelected` tries to achieve. IIRC, it is a heuristic to determine whether the reference is _the_ reference intended to be selected by the user. The heuristic currently in place is that the user wants to select the reference at the start of whatever selection he does.

The best selection determined this way is used by the commend bar and refactorings like rename.

In short, as far as I know the goal of `IsSelected` is not to determine whether a reference is encompassed in the current selection. To do that, you have to get the current selection and Chechnya whether it contains the selection of the reference.

**Comment by Vogel612** [comment]
I'm not quite clear on your final goal here, but a way to check whether one selection **contains** another would be the following:

```
public bool Contains(Selection other) 
{
    bool beginsAfterStart = (StartLine == other.StartLine && StartColumn <= other.StartColumn) || StartLine < other.StartLine;
    bool endsBeforeEnd = (EndLine == other.EndLine && EndColumn >= other.EndColumn) || EndLine > other.EndLine;
    return beginsAfterStart && endsBeforeEnd;
}
```

**Comment by Vogel612** [comment]
@MDoerner sounds like a good place to drop an XML-Doc

**Comment by MDoerner** [comment]
Moreover, the original logic of `ContainsFirstCharacter` seems to be completely fine. The goal is to determine whether the start of the selection passed in is contained in the selection on which this member is called. The change above makes it equivalent to `Contains`.

**Comment by bclothier** [comment]
@MDoerner,

Regarding the `ContainsFirstCharacter`, I'm not sure everything's fine. I expected the function to use a new selection of `L9C17`, not `L9C1` as I was getting in the original code, because of the indentation which are just whitespace. I changed that as a test to see if that was part of problem but as noted that only created more problems than solutions. Also, it seemed that the process is reverse - it should be the reference's selection collapsing to `L9C17` and then tested in `Contains` against the user's selection of `L9C1-L10C21`. But what actually happens is user's selection gets collapsed to `L9C1` and compared to reference's selection of `L917-L9C21`, which doesn't make much sense given the name of the function. 

I do agree with @Vogel612  about XML-Docs since it wasn't obvious to me that `IsSelected` isn't going to return true for an encompassing selection. I also am not sure if it really ought to return false because if it's encompassed, it *is* selected, no? Furthermore, if it's intended to be exclusive to that reference, I'm not sure we are actually checking whether the selection is exclusive to that reference because we are simply collapsing the selection so we aren't checking if the selection doesn't contain other reference.

**Comment by MDoerner** [comment]
@bclothier `ContainsFirstCharacter` is working perfectly fine. Please have a look again at `IsSelected`. There, `ContainsFirstCharacter` gets called on the selection of the reference with the selection from the qualified selection as argument. Thus, it correctly determines that the first character of the selection passed in is not contained in the selection of the reference.

**Comment by MDoerner** [comment]
Regarding the heuristic to determine _the_ reference the user wanted to select: the base problem is that there must be at most one reference marked as _the_ selected one because that is what the command bar and the refactorings need. 
Moreover, I did not invent this heuristic. There might be better ways to deal with this, like taking the first reference that intersects the selection. However, one would need to put some thought into it in order to avoid strange bugs.

**Comment by bclothier** [comment]
Just so it's clear, I reverted my local copy of `Selection.cs` and used a similar solution, using the `Selection.Contains` directly. If someone can put in a PR for adding XML-Docs to those and clarify the intention, I think that'll close the issue because if I can misinterpret the intention of the `IsSelected`, others may.

**Comment by bclothier** [comment]
Closing since question's been answered
